accepted/2024/supplied-by-platform.md

```diff
@@ -0,0 +1,192 @@
+
+- Pros:
+  - Performance benefits.
+  These lists will contain hundreds of packages.
+  NuGet can cache the package list based on a file. With the current implementation, NuGet needs to maintain a list per framework, per project, arguably with a lot of duplication. Even if NuGet deduplicates the objects reprsenting the packages in some way, it would still register from a memory allocations perspective(which has been a focus), on both NuGet & project-system side.
+  - Eliminate potential misuse.
+- Cons:
+  - The feature is no longer easy to generalize.
+  There are various asks for the ability to simply ignore certain packages from graphs. More often than not, the reason is that an old package is being referenced and cannot be updated.
+
+### Not doing the feature
+
+With security being a focus of NuGet as a package manager, minimizing false positives is essential. The fact that there are still assemblies that ship both as packages and as part of shared frameworks means that the changes of false positives are not going away with a framework update.
+We can consider not doing this feature and relying on platform packages to simply go away, especially with NuGetAudit enabling transitive dependency auditing recently, but that is not going to solve the long term concern of assemblies such `Microsoft.Extensions.Logging`.
+
+## Prior Art
+
+<!-- What prior art, both good and bad are related to this proposal? -->
+<!-- Do other features exist in other ecosystems and what experience have their community had? -->
+<!-- What lessons from other communities can we learn from? -->
+<!-- Are there any resources that are relevant to this proposal? -->
+
+- This problem is .NET specific.
+
+## Unresolved Questions
+
+<!-- What parts of the proposal do you expect to resolve before this gets accepted? -->
+<!-- What parts of the proposal need to be resolved before the proposal is stabilized? -->
+<!-- What related issues would you consider out of scope for this proposal but can be addressed in the future? -->
+
+- MSBuild items/properties vs a file with the data
+- MSBuild item name ideas: `PrunedPackageReference` was used. `IgnoredPackageReference` and `SkippedPackageReference` are alternatives.
+- What if a project id is specified in PrunedPackageReference? Warn? Error? Skip? Prune & Warn?
+- What to do when a direct PackageReference is specified to be pruned?
+- Should the pruned package reference dissapear from the dependencies section completely? Strong preference towards no, since it aids visibility.
+- Should the assets file contain the list of pruned packages? Are only ids important, or do we need versions as well?
+- Should the version attribute of PrunedPackageReference be a version range instead? Should the attribute be named MaxVersion instead?
+- What's the list of packages that'll be removed by the .NET SDK? How do we verify it won't break things?
```

Comment:
When referencing the Microsoft.NETCore.App 9.0 shared framework, the list of packages to prune is https://github.com/dotnet/runtime/blob/main/src/installer/pkg/sfx/Microsoft.NETCore.App/PackageOverrides.txt

When referencing the Microsoft.AspNetCore.App 9.0 shared framework, the list of packages is https://github.com/dotnet/aspnetcore/blob/main/eng/PackageOverrides.txt

Looks like there isn't a PackageOverrides.txt file when referencing the Microsoft.WindowsDesktop.App shared framework. We probably should add one.

Same for older frameworks. The PackageOverrides list is part of the shared framework targeting pack and the SDK will need that information before restore.

Reply:
Those are the ones that I used as dummy values in my tests. https://github.com/NuGet/NuGet.Client/blob/f4519cb5fa25de9080ce2a66ff580354c61c9c3e/src/NuGet.Core/NuGet.Build.Tasks/PopulatePrunedPackageReferencesTask.cs#L37. 

Given what you said about shared frameworks, I think we have  a challenge with the transitive framework references. 
In particular, a package can specify a shared framework even if the project itself doesn't require it. 
Pruning that would be challenging, since the .NET SDK cannot know about the framework refs brought through by packages. 

Pruning is a graph walk time feature. Transitive framework references are known after the graph walk has been completed. 

We don't have an easy solution (if any) to this particular problem. In theory, I think the .NET SDK can actually provide switches for pruning additional shared frameworks. However, there's the potential pitfall where pruning those packages may lead to the shared framework being pruned as well. It's detectable by the SDK at build time though. 

This also brings up a very important thing where the .NET SDK should only prune frameworks that are directly specified to avoid first and second restore being different.